crypto/rsa/rsa_gen.c

```diff
@@ -31,16 +31,19 @@ static int rsa_builtin_keygen(RSA *rsa, int bits, int primes, BIGNUM *e_value,
  */
 int RSA_generate_key_ex(RSA *rsa, int bits, BIGNUM *e_value, BN_GENCB *cb)
 {
-    return RSA_generate_multi_prime_key(rsa, bits, RSA_DEFAULT_PRIME_NUM,
-                                        e_value, cb);
+    if (rsa->meth->rsa_keygen != NULL)
+        return rsa->meth->rsa_keygen(rsa, bits, e_value, cb);
+
+    return rsa_builtin_keygen(rsa, bits, RSA_DEFAULT_PRIME_NUM, e_value, cb);
 }
 
 int RSA_generate_multi_prime_key(RSA *rsa, int bits, int primes,
                                  BIGNUM *e_value, BN_GENCB *cb)
 {
     /* multi-prime is only supported with the builtin key generation */
-    if (rsa->meth->rsa_keygen != NULL)
-        return rsa->meth->rsa_keygen(rsa, bits, e_value, cb);
+    if (rsa->meth->rsa_multi_prime_keygen != NULL)
+        return rsa->meth->rsa_multi_prime_keygen(rsa, primes,
+                                                 bits, e_value, cb);
```

Comment:
"bits, primes" not "primes, bits" if you honour prototype. Or vice versa... I mean maybe prototype needs to be fixed.

Reply:
Good catch...

Reply:
fixed

Reply:
On side note related to my comment elsewhere regarding branch optimizations. Not that it's significant here (since RSA operations are computationally heavy), but processors would "feel" better if it was `if ( ... == NULL) rsa_bultin_keygen()`. If branch is not in branch prediction "cache", processor will assume it's not taken here, and as it stands now will speculate to indirect call. But indirect calls/branches are impossible to predict, and hence there is no penalty-free path...